tests/test_ui_server.py

```diff
@@ -34,3 +42,101 @@ def test_get_vector_colnames(self):
         colnames = json.loads(response.get_data(as_text=True))
         self.assertEqual(response.status_code, 422)
         self.assertEqual(colnames, [])
+
+    def test_get_invest_models(self):
+        """UI server: get_invest_models endpoint."""
+        from natcap.invest import ui_server
+        test_client = ui_server.app.test_client()
+        response = test_client.get('/models')
+        models_dict = json.loads(response.get_data(as_text=True))
+        for model in models_dict.values():
+            self.assertEqual(list(model), ['internal_name', 'aliases'])
+
+    def test_get_invest_spec(self):
+        """UI server: get_invest_spec endpoint."""
+        from natcap.invest import ui_server
+        test_client = ui_server.app.test_client()
+        response = test_client.post('/getspec', json='sdr')
+        spec = json.loads(response.get_data(as_text=True))
+        self.assertEqual(
+            list(spec),
+            ['model_name', 'module', 'userguide_html',
+             'args_with_spatial_overlap', 'args'])
+
+    def test_get_invest_validate(self):
+        """UI server: get_invest_validate endpoint."""
+        from natcap.invest import ui_server, carbon
+        test_client = ui_server.app.test_client()
+        args = {
+            'workspace_dir': 'foo'
+        }
+        payload = {
+            'model_module': carbon.ARGS_SPEC['module'],
+            'args': json.dumps(args)
+        }
+        response = test_client.post('/validate', json=payload)
+        results = json.loads(response.get_data(as_text=True))
+        expected = carbon.validate(args)
+        # These differ only because a tuple was transformed to a list during
+        # the json (de)serializing, so do the same with expected data
+        self.assertEqual(results, json.loads(json.dumps(expected)))
+
+    def test_post_datastack_file(self):
+        """UI server: post_datastack_file endpoint."""
+        from natcap.invest import ui_server
+        test_client = ui_server.app.test_client()
+        self.workspace_dir = tempfile.mkdtemp()
+        expected_datastack = {
+            'args': {
+                'workspace_dir': 'foo'
+            },
+            'invest_version': '3.10.0',
+            'model_name': 'natcap.invest.carbon'
+        }
+        filepath = os.path.join(self.workspace_dir, 'datastack.json')
+        with open(filepath, 'w') as file:
+            file.write(json.dumps(expected_datastack))
+        response = test_client.post('/post_datastack_file', json=filepath)
+        response_data = json.loads(response.get_data(as_text=True))
+        self.assertEqual(
+            list(response_data),
+            ['type', 'args', 'module_name', 'model_run_name',
```

Comment:
It looks like the `json` module preserves the order of keys, so this should work. But JSON is generally unordered outside of the python implementation, so I think `ui_server.post_datastack_file` could put them in any order without breaking its contract. Maybe use a set here just in case?

Reply:
Great point! I changed all these tests to use sets instead of lists.